tests/src/Clash/Vivado.hs

```diff
@@ -0,0 +1,149 @@
+                       } deriving Show
+
+instance Semigroup SimProj where
+  (<>) (SimProj vhd0 v0 ip0) (SimProj vhd1 v1 ip1) =
+    SimProj (vhd0<>vhd1) (v0<>v1) (ip0<>ip1)
+
+instance Monoid SimProj where
+  mempty = SimProj [] [] []
+
+thread :: [a -> a] -> a -> a
+thread = foldr (.) id
+
+-- | Given something like @topEntity_dcfifo_240A2A9D3FF8D64B.tcl@, return the
+-- IP name, @dcfifo@
+parseTcl :: FilePath -> String
+parseTcl fp = takeWhile (/= '_') (drop 1 $ dropWhile (/= '_') fp)
+
+mSimProj :: FilePath -> T.Text -> FilePath -> SimProj -> SimProj
+mSimProj dir lib fp p@(SimProj vhd v ip)
+  | ".vhdl" `isSuffixOf` fp = SimProj ((lib, dir </> fp):vhd) v ip
+  | ".vhd" `isSuffixOf` fp = SimProj ((lib, dir </> fp):vhd) v ip
+  | ".v" `isSuffixOf` fp = SimProj vhd ((lib, dir </> fp):v) ip
+  | ".tcl" `isSuffixOf` fp = SimProj vhd v ((dir </> fp):ip)
+  | otherwise = p
+
+simProjFromClashEntities ::
+  -- | HDL source dir
+  FilePath ->
+  -- | Qualified name
+  String ->
+  IO SimProj
+simProjFromClashEntities hdlDir qualName = do
+  (Just mHdl) <- readManifest (hdlDir </> qualName ++ "/clash-manifest.json")
+  let deps = T.unpack <$> transitiveDependencies mHdl
+  nextSimProj <- traverse (simProjFromClashEntities hdlDir) deps
+  pure $ mconcat nextSimProj <> asSimProj qualName mHdl
+
+fromModuleName ::
+  HdlSource ->
+  -- | Module name
+  String ->
+  -- | Entity
+  String ->
+  -- | TCL script
+  IO String
+fromModuleName src moduleName entity = do
+  proj <- simProjFromClashEntities (directory src) (moduleName ++ "." ++ entity)
+  cabalDir <- getDataDir
+  pure $ mkTcl cabalDir entity src proj
+
+asSimProj :: FilePath -> Manifest -> SimProj
+asSimProj dir m =
+  let fps = fst <$> fileNames m
+    in thread (mSimProj dir (topComponent m) <$> fps) mempty
+
+mkTcl ::
+  FilePath ->
+  -- | Cabal @data@ dir
+  String ->
+  -- | Entity
+  HdlSource ->
+  SimProj ->
+  String
+mkTcl cabalDir entity hdl (SimProj vhdls vs tcls) =
+  [i|
+source #{cabalDir}/tcl/clash_namespace.tcl
+
+# create clash namespace
+namespace eval clash {}
+
+set clash::hdl_dir #{hdlDir}
+
+# for cleanup
+file delete -force xilinx_tmp
+file mkdir xilinx_tmp
+cd xilinx_tmp
+
+create_project -in_memory
+
+set clash::ips [list]
+|]
+
+  ++ concatMap (\tcl ->
+      [i|clash::loadTclIface #{parseTcl tcl} ${clash::hdl_dir}/#{tcl}\n|]) tcls
+
+  ++
+      [i|
+
+set clash::ips [clash::runClashScripts]
+set clash::ipFiles [get_property IP_FILE [get_ips $clash::ips]]
+
+read_ip $clash::ipFiles
+
+generate_target {synthesis simulation} [get_ips $clash::ips]
+
+foreach modName $clash::ips {
+  add_files .gen/sources_1/ip/${modName}/synth
+}
+|]
+
+  ++ readSources
+
+  ++ [i|
+update_compile_order -fileset [current_fileset]
+
+set_property top #{entity} [get_fileset sim_1]
+
+save_project_as sim -force
+
+launch_simulation -simset sim_1 -mode behavioral
+
+run -all
+
+quit
+|]
+ where
+  hdlDir = directory hdl
+  readSources = case buildTarget hdl of
```

Comment:
Oh yes I see.

Reply:
This is just how I organized the code, I don't think it's much better doing a `case` on the extension, but I don't think there's much advantage either way.

Reply:
It looks rather convoluted to me.

First you `case` on the extension, to put it in two separate lists `vhdlFiles` and `verilogFiles`. Then you keep the `buildTarget` around just for these lines of code, meaning you need the datatype `HdlSource` to put it in, whereas otherwise you could have just kept `directory` from that record type. Then you `case` on the `buildTarget` to pick the right one of the two lists you built by casing on the extension. Why not just `case` on the extension once, here?

Reply:
Ah right, no, you do need to `case` on the extension to pick out the `.tcl` earlier, I overlooked that for a moment. But then you could just add all files in `vhdlFiles` with `read_vhdl` and add all files from `verilogFiles` with `read_verilog`, again eliminating the need to pass `buildTarget` around and `case` on it here. In practice, one of those will always be empty, unless we suddenly begin generating `.vhdl` and `.v` in the same project. I don't see a reason for it now, but I also don't see a reason to filter it out (which to me is the only purpose of this code, is there another?).

Reply:
Ok, now I no longer split into VHDL files vs. Verilog files but rather assume based on the project type. I think this is fine since all the tests specify one HDL or the other.